File: nn/activation.go
package nn

import "math"

var ActivationMap = map[string]func(float64) float64{
	"sigmoid": Sigmoid,
	"tanh":    Tanh,
	"relu":    Relu,
}

func Sigmoid(x float64) float64 {
	return 1.0 / (1.0 + math.Exp(-x))
}

func SigmoidPrime(x float64) float64 {
	s := Sigmoid(x)
	return s / (1.0 - s)
}

func Tanh(x float64) float64 {
	return math.Tanh(x)
}

func TanhPrime(x float64) float64 {
	return math.Pow((1.0 / math.Cosh(x)), 2)
}

func Relu(x float64) float64 {
	if x > 0 {
		return x
	}
	return 0
}

func ReluPrime(x float64) float64 {
	// maybe want to look into edge case if x == 0
	if x > 0 {
		return 1
	}
	return 0
}

File: nn/activation.py
from typing import Callable
from nn.nn import NN
import numpy as np


def get_activation(nn: NN) -> Callable:
    a = nn.activation
    funcs = {
        "relu": relu,
        "sigmoid": sigmoid,
        "tanh": tanh,
    }

    prime_funcs = {
        "sigmoid": sigmoid_prime,
        "tanh": tanh_prime,
        "relu": relu_prime,
    }

    nn.set_func(funcs[a])
    nn.set_func_prime(prime_funcs[a])


def relu(x):
    return np.maximum(0.0, x)


def relu_prime(x):
    return np.maximum(0, x)


def sigmoid(x):
    return 1.0 / (1.0 + np.exp(-x))


def sigmoid_prime(x):
    s = sigmoid(x)
    return s * (1 - s)


def tanh(x):
    return np.tanh(x)


def tanh_prime(x):
    return 1 - np.tanh(x)**2
